pkg/server/home_workspaces.go

```diff
@@ -0,0 +1,661 @@
+				responsewriters.InternalError(rw, req, err)
+				return
+			} else if decision != authorizer.DecisionAllow {
+				responsewriters.Forbidden(ctx, attributes, rw, req, reason, homeWorkspaceCodecs)
+				return
+			}
+
+			var createError error
+			retryLater, retryAfterSeconds, createError = h.tryToCreate(ctx, user.GetName(), workspaceName, workspaceType)
+			if createError != nil {
+				responsewriters.ErrorNegotiated(createError, errorCodecs, schema.GroupVersion{}, rw, req)
+				return
+			}
+		}
+
+		if retryLater {
+			// Return a 429 status asking the client to try again after the creationDelay
+			rw.Header().Set("Retry-After", fmt.Sprintf("%d", retryAfterSeconds))
+			http.Error(rw, "Creating the home workspace", http.StatusTooManyRequests)
+			return
+		}
+	}
+	h.apiHandler.ServeHTTP(rw, req)
+}
+
+var reHomeWorkspaceChar = regexp.MustCompile("[^a-z0-9-]")
+
+// getHomeWorkspaceName returns the full name of the home workspace for a given user
+// This full name includes home bucket workspaces whose name is based on the user name
+// sha1 hash.
+func (h *homeWorkspaceHandler) getHomeWorkspaceName(userName string) logicalcluster.Name {
+	// bucketLevels <= 5
+	// bucketSize <= 4
+	bytes := sha1.Sum([]byte(userName))
+
+	result := h.homePrefix
+	for level := 0; level < h.bucketLevels; level++ {
+		var bucketBytes = make([]byte, h.bucketSize)
+		bucketBytesStart := level
+		bucketCharInteger := binary.BigEndian.Uint32(bytes[bucketBytesStart : bucketBytesStart+4])
+		for bucketCharIndex := 0; bucketCharIndex < h.bucketSize; bucketCharIndex++ {
+			bucketChar := byte('a') + byte(bucketCharInteger%26)
+			bucketBytes[bucketCharIndex] = bucketChar
+			bucketCharInteger /= 26
+		}
+		result = result.Join(string(bucketBytes))
+	}
+
+	userName = reHomeWorkspaceChar.ReplaceAllLiteralString(userName, "-")
+	userName = strings.TrimLeftFunc(userName, func(r rune) bool {
+		return r <= '9'
+	})
+	userName = strings.TrimRightFunc(userName, func(r rune) bool {
+		return r == '-'
+	})
+
+	return result.Join(userName)
+}
+
+// needsAutomaticCreation deduces, from the workspace full name,
+// according the the expected home root and home bucket level number,
+// whether this workspace has to be checked for automatic creation
+// and what its workspace type will be.
+func (h *homeWorkspaceHandler) needsAutomaticCreation(workspaceName logicalcluster.Name) (needsCheck bool, workspaceType tenancyv1alpha1.ClusterWorkspaceTypeName) {
+	if !workspaceName.HasPrefix(h.homePrefix) {
+		return
+	}
+	length := strings.Count(workspaceName.String(), ":")
+	homeRootWorkspaceLength := strings.Count(h.homePrefix.String(), ":")
+	lastBucketWorkspaceLength := homeRootWorkspaceLength + h.bucketLevels
+	homeWorkspaceLength := lastBucketWorkspaceLength + 1
+	switch {
+	case length <= homeRootWorkspaceLength:
+		return
+	case length <= lastBucketWorkspaceLength:
+		return true, "Homebucket"
+	case length == homeWorkspaceLength:
+		return true, "Home"
+	}
+	return
+}
```

Comment:
this function is fragile. Please only use method of the logicalcluster. It has everything necessary here.